src/test/socket/bind/test_bind.rs

```diff
@@ -177,9 +178,112 @@ fn get_tests() -> Vec<test_utils::ShadowTest<(), String>> {
         }
     }
 
+    // additional tests
+    tests.push(test_utils::ShadowTest::new(
+        "tcp-reuse-addr-with-orphaned-child-socket",
+        test_tcp_reuse_addr_with_orphaned_child_socket,
+        set![TestEnv::Libc, TestEnv::Shadow],
+    ));
+
     tests
 }
 
+fn test_tcp_reuse_addr_with_orphaned_child_socket() -> Result<(), String> {
+    // This is essentially a regression test for
+    // https://github.com/shadow/shadow/issues/3563.
+    // We bind a socket, let a client connect, close the binding socket,
+    // and then try to bind a new socket to the same address.
+
+    let addr = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 11111u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_LOOPBACK.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+    let domain = libc::AF_INET;
+    let sock_type = libc::SOCK_STREAM;
+    let socket_flag = 0;
+    let listen_fd = unsafe { libc::socket(domain, sock_type | socket_flag, 0) };
+    assert_with_errno!(listen_fd >= 0);
+
+    // We set SO_REUSEADDR to avoid having to wait for TIME_WAIT before allowing
+    // the address to be reused. This currently isn't strictly needed under
+    // shadow, since the `close` syscall immediately fully closes the listening
+    // socket without waiting TIME_WAIT (possibly a bug?).
```

Comment:
I think that the listening socket on Linux doesn't ever enter the TIME_WAIT state. This doesn't follow the wording of RFC 9293, but \*shrug emoji\*. (See https://github.com/shadow/shadow/blob/main/src/lib/tcp/notes/listener-close-with-pending-conn.md)

My understanding on Linux is that what can happen is that any previously accepted sockets can go to TIME_WAIT, which prevents new listening sockets from binding to the same address if the socket wasn't originally bound with `SO_REUSEADDR`. So it's not the (closed) listening socket itself that prevents new sockets from reusing the same address, but previously accepted sockets that are now in the TIME_WAIT state.